internal/terraform/node_action_validate.go

```diff
@@ -82,16 +82,28 @@ func (n *NodeValidatableAction) Execute(ctx EvalContext, _ walkOperation) tfdiag
 		return diags
 	}
 
-	var configVal cty.Value
-	var valDiags tfdiags.Diagnostics
-	if n.Config.Config != nil {
-		configVal, _, valDiags = ctx.EvaluateBlock(n.Config.Config, schema.ConfigSchema, nil, keyData)
-		diags = diags.Append(valDiags)
-		if valDiags.HasErrors() {
+	config := n.Config.Config
+	if n.Config.Config == nil {
+		config = hcl.EmptyBody()
+	}
+
+	configVal, _, valDiags := ctx.EvaluateBlock(config, schema.ConfigSchema, nil, keyData)
+	if valDiags.HasErrors() {
+		// If there was no config block at all, we'll build a more-detailed diagnostic with Subject range
+		if n.Config.Config == nil {
+			for _, diag := range valDiags {
```

Comment:
the result would be the same, but there is also `Diagnostic.ToHCL` which does this conversion for you, and then you could just inject the source location within the loop.